src/models/geograph.py

```diff
@@ -66,50 +305,135 @@ def _dataframe_to_graph(
         Args:
             df (gpd.GeoDataFrame): GeoDataFrame containing polygon objects from
             a shape file.
-            attributes (Optional[List[str]], optional): columns of gdf that are
-            added to nodes of graph as attributes. Defaults to None.
+            attributes (Optional[List[str]], optional): columns of df that are
+            added to nodes of graph as attributes. Defaults to None,
+            which will cause all attributes to be added.
+            tolerance (float, optional): Adds edges between neighbours that are
+            at most `tolerance` metres apart. Defaults to 0.
+
+        Raises:
+            ValueError: If `tolerance` < 0.
+
+        Returns:
+            STRtree: The Rtree object used to build the graph.
         """
+        if tolerance < 0.0:
+            raise ValueError("`tolerance` must be greater than 0.")
         # If no attribute list is given, all
         # columns in df are used
         if attributes is None:
             attributes = df.columns.tolist()
+        elif not set(attributes).issubset(set(df.columns)):
+            raise ValueError("`attributes` must only contain column names in `df`.")
 
-        geom = df.geometry.tolist()
+        geom: Dict[int, shapely.Polygon] = df.geometry.to_dict()
         # Build dict mapping hashable unique object ids for each polygon
         # to their index in geom
-        id_dict = {id(polygon): index for index, polygon in enumerate(geom)}
+        id_dict: Dict[int, int] = {
+            id(polygon): index for index, polygon in geom.items()
+        }
         # Build Rtree from geometry
-        tree = STRtree(geom)
+        tree = STRtree(geom.values())
         # this dict maps polygon indices in df to a list
         # of neighbouring polygon indices
         graph_dict = {}
 
         # Creating nodes (=vertices) and finding neighbors
         for index, polygon in tqdm(
-            enumerate(geom),
+            geom.items(),
             desc="Step 1 of 2: Creating nodes and finding neighbours",
             total=len(geom),
         ):
-            # find the indexes of all polygons which touch the borders of or
-            # overlap with this one
-            neighbours = [
-                id_dict[id(nbr)]
-                for nbr in tree.query(polygon)
-                if nbr.touches(polygon) or nbr.overlaps(polygon)
-            ]
+            if tolerance > 0:
```

Comment:
Sweet! Just out of curiosity, is this still 6x slower or was there a speed-up versus the `distance` based approach you mentioned this morning?

Reply:
I tested the `distance` based approach fully and it was actually ~13x slower, but this approach is only a fraction slower, maybe it takes ~20% more time. This makes sense since the only extra computation here is the `buffer` call.

Reply:
Cool, thank you for the extra info! Great tests!